clientconn.go

```diff
@@ -1135,112 +1158,86 @@ func (ac *addrConn) adjustParams(r transport.GoAwayReason) {
 }
 
 func (ac *addrConn) resetTransport() {
-	for i := 0; ; i++ {
-		if i > 0 {
-			ac.cc.resolveNow(resolver.ResolveNowOptions{})
-		}
+	ac.mu.Lock()
+	if ac.state == connectivity.Shutdown {
+		ac.mu.Unlock()
+		return
+	}
+
+	addrs := ac.addrs
+	backoffFor := ac.dopts.bs.Backoff(ac.backoffIdx)
+	// This will be the duration that dial gets to finish.
+	dialDuration := minConnectTimeout
+	if ac.dopts.minConnectTimeout != nil {
+		dialDuration = ac.dopts.minConnectTimeout()
+	}
+
+	if dialDuration < backoffFor {
+		// Give dial more time as we keep failing to connect.
+		dialDuration = backoffFor
+	}
+	// We can potentially spend all the time trying the first address, and
+	// if the server accepts the connection and then hangs, the following
+	// addresses will never be tried.
+	//
+	// The spec doesn't mention what should be done for multiple addresses.
+	// https://github.com/grpc/grpc/blob/master/doc/connection-backoff.md#proposed-backoff-algorithm
+	connectDeadline := time.Now().Add(dialDuration)
+
+	ac.updateConnectivityState(connectivity.Connecting, nil)
+	ac.mu.Unlock()
 
+	if err := ac.tryAllAddrs(addrs, connectDeadline); err != nil {
+		ac.cc.resolveNow(resolver.ResolveNowOptions{})
+		// After exhausting all addresses, the addrConn enters
+		// TRANSIENT_FAILURE.
 		ac.mu.Lock()
 		if ac.state == connectivity.Shutdown {
 			ac.mu.Unlock()
 			return
 		}
+		ac.updateConnectivityState(connectivity.TransientFailure, err)
 
-		addrs := ac.addrs
-		backoffFor := ac.dopts.bs.Backoff(ac.backoffIdx)
-		// This will be the duration that dial gets to finish.
-		dialDuration := minConnectTimeout
-		if ac.dopts.minConnectTimeout != nil {
-			dialDuration = ac.dopts.minConnectTimeout()
-		}
-
-		if dialDuration < backoffFor {
-			// Give dial more time as we keep failing to connect.
-			dialDuration = backoffFor
-		}
-		// We can potentially spend all the time trying the first address, and
-		// if the server accepts the connection and then hangs, the following
-		// addresses will never be tried.
-		//
-		// The spec doesn't mention what should be done for multiple addresses.
-		// https://github.com/grpc/grpc/blob/master/doc/connection-backoff.md#proposed-backoff-algorithm
-		connectDeadline := time.Now().Add(dialDuration)
-
-		ac.updateConnectivityState(connectivity.Connecting, nil)
-		ac.transport = nil
+		// Backoff.
+		b := ac.resetBackoff
 		ac.mu.Unlock()
 
-		newTr, addr, reconnect, err := ac.tryAllAddrs(addrs, connectDeadline)
-		if err != nil {
-			// After exhausting all addresses, the addrConn enters
-			// TRANSIENT_FAILURE.
+		timer := time.NewTimer(backoffFor)
```

Comment:
As per https://github.com/grpc/grpc/blob/master/doc/connectivity-semantics-and-api.md#states-of-connectivity, the condition to move from `CONNECTING` to `IDLE` is:
```
No RPC activity on channel for IDLE_TIMEOUT
```

What is `IDLE_TIMEOUT`? We don't have a dialOption for that. Do other languages have?

But here, we have already moved from `CONNECTING` to `TRANSIENT_FAILURE` and there doesn't seem to be a valid transition from `TRANSIENT_FAILURE` to `IDLE` as per the gRFC. Not sure I quite follow what is going on here.

Also, the gRFC talks about moving from `READY` to `IDLE`. Are we not planning on support that?

Reply:
> What is IDLE_TIMEOUT? We don't have a dialOption for that. Do other languages have?

We don't support channel idleness, which is what IDLE_TIMEOUT is about.  It tears down the whole channel, basically (name resolver included), but leaves it in a state that can be restarted.  Only Java supports it, I'm pretty sure.

That doc is not comprehensive, and there are some subtle differences between the Go/Java model of subchannels and the C model, which the document describes.  This was discussed pretty extensively and we agreed this was better than going back to connecting directly, since there may be no pending RPCs.

> Also, the gRFC talks about moving from READY to IDLE. Are we not planning on support that?

This PR implements that.  It happens when the transport is closed, in the callbacks in `createTransport`.

Reply:
Ack.